File: commons/org.beangle.commons.core/src/main/scala/org/beangle/commons/i18n/AbstractTextResource.scala
/*
 * Beangle, Agile Java/Scala Development Scaffold and Toolkit
 *
 * Copyright (c) 2005-2012, Beangle Software.
 *
 * Beangle is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Beangle is distributed in the hope that it will be useful.
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public License
 * along with Beangle.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.beangle.commons.i18n

import java.text.MessageFormat

abstract class AbstractTextResource extends TextResource {

  def get(key: String, defaultValue: String, args: Any*): String = {
    val text = getText(key).getOrElse(defaultValue);
    new MessageFormat(text, locale).format(args)
  }

  protected def getText(key: String): Option[String];
}

import java.util.Locale

class NullTextResource extends AbstractTextResource {
  override def locale = Locale.getDefault

  override def getText(key: String) = Some(key)
}

import java.util.ResourceBundle

class BundleTextResource(val locale: Locale, val bundle: ResourceBundle) extends AbstractTextResource {
  override def getText(key: String): Option[String] = {
    val text = bundle.getString(key)
    if (null == text) None else Some(text)
  }
}

File: commons/org.beangle.commons.core/src/main/java/org/beangle/commons/i18n/AbstractTextResource.java
/* Copyright c 2005-2012.
 * Licensed under GNU  LESSER General Public License, Version 3.
 * http://www.gnu.org/licenses
 */
package org.beangle.commons.i18n;

import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;

/**
 * <p>
 * Abstract AbstractTextResource class.
 * </p>
 * 
 * @author chaostone
 * @version $Id: $
 */
public abstract class AbstractTextResource implements TextResource{

  /** Constant <code>EMPTY_ARGS</code> */
  protected static final Object[] EMPTY_ARGS = new Object[0];

  /**
   * <p>
   * getText.
   * </p>
   * 
   * @param key a {@link java.lang.String} object.
   * @param defaultValue a {@link java.lang.String} object.
   * @return a {@link java.lang.String} object.
   */
  public String getText(String key, String defaultValue) {
    String text = getText(key);
    if (null == text) { return defaultValue; }
    return text;
  }

  /**
   * <p>
   * getText.
   * </p>
   * 
   * @param key a {@link java.lang.String} object.
   * @param defaultValue a {@link java.lang.String} object.
   * @param obj a {@link java.lang.String} object.
   * @return a {@link java.lang.String} object.
   */
  public String getText(String key, String defaultValue, String obj) {
    List<Object> args = new ArrayList<Object>(1);
    args.add(obj);
    return getText(key, defaultValue, args);
  }

  /**
   * <p>
   * getText.
   * </p>
   * 
   * @param key a {@link java.lang.String} object.
   * @param args a {@link java.util.List} object.
   * @return a {@link java.lang.String} object.
   */
  public String getText(String key, List<Object> args) {
    Object[] params;
    if (null == args) {
      params = EMPTY_ARGS;
    } else {
      params = args.toArray();
    }
    return getText(key, params);
  }

  /** {@inheritDoc} */
  public String getText(String key, String defaultValue, List<Object> args) {
    Object[] params;
    if (null == args) {
      params = EMPTY_ARGS;
    } else {
      params = args.toArray();
    }
    return getText(key, defaultValue, params);
  }

  /**
   * <p>
   * getText.
   * </p>
   * 
   * @param key a {@link java.lang.String} object.
   * @param defaultValue a {@link java.lang.String} object.
   * @param args an array of {@link java.lang.Object} objects.
   * @return a {@link java.lang.String} object.
   */
  public String getText(String key, String defaultValue, Object[] args) {
    String text = getText(key, args);
    String defaultVal = defaultValue;
    if (null == text && null == defaultVal) {
      defaultVal = key;
    }
    if (null == text && defaultVal != null) {
      MessageFormat format = new MessageFormat(defaultVal);
      format.setLocale(getLocale());
      format.applyPattern(defaultVal);
      return format.format(args);
    }
    return text;
  }

}
